### PR TITLE
gRPC: fix cases where gRPC call could be finished twice

### PR DESCRIPTION
**Pull request by var-const**
gRPC C++ wrappers assert that performing any operations on a call results in an "ok" error code. In certain cases, it was possible for Firestore to call `grpc::ClientAsyncReaderWriter::Finish` twice, which causes that assertion to fail, crashing the app. I haven't succeeded in reproducing the issue, but I presume the following would be consistent with bug reports:
* in quick succession, an operation fails, and then `Stream` stops due to idleness. Normally, a failed operation would cause the stream to stop, and also when an operation fails, `Stream` doesn't call `FinishImmediately` on the `GrpcStream`. However, handling operation failure is split into two async steps: the first one is to finish the underlying gRPC call with a completion, and the second is to invoke the completion which will notify the parent `Stream`. Because the second step gets added to the back of the worker queue, theoretically the following configuration is possible:
```
[GrpcStream::OnOperationFailed] [...] [...] [Stream::Stop() /* idleness */] [...] [GprcStream::FinishAndNotify]
                        v -----------------------------------------------------------------------^
```
Stop due to idleness is considered a graceful stop, and so will call `GrpcStream::FinishImmediately`. `GrpcStream::Shutdown` relies on `completions_.empty()` check to avoid double finish. However, in this case, the check is insufficient, because `GrpcStream` still has one completion -- the "finish" operation enqueued in `OnOperationFailed` that won't be removed until `FinishAndNotify`.
* if `GrpcStream::OnOperationFailed` is called twice in quick succession, so that the second `OnOperationFailed` is executed before `FinishAndNotify` enqueued by the first `OnOperationFailed` is. The failure is similar to case 1.

These sound like pretty rare circumstances, which is consistent with how difficult the issue is to reproduce. There might be more cases; the best fix seems to be ensuring that gRPC call is never finished twice.

This PR:
* adds a flag, `is_grpc_call_finished_`, to track whether the underlying call has been finished without trying to interpret the current state of `GrpcStream` to infer that information;
* adds logging to the shutdown sequence of `GrpcStream`, in case more related crashes are discovered;
* adds tests for the two situations described above. Manually verified that none of the tests pass on master.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/remote/grpc_stream.cc [hunk 1]
```diff
@@ -226,30 +246,35 @@ void GrpcStream::FastFinishCompletionsBlocking() {
 
 bool GrpcStream::WriteAndFinish(grpc::ByteBuffer&& message) {
   bool did_last_write = false;
+  if (!is_grpc_call_finished_) {
+    did_last_write = TryLastWrite(std::move(message));
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/grpc_stream.cc]: This refactoring is to make the handling of `is_grpc_call_finished_` in this case more explicit.

File: Firestore/core/src/firebase/firestore/remote/grpc_stream.cc [hunk 2]
```diff
@@ -173,22 +179,28 @@ void GrpcStream::FinishAndNotify(const Status& status) {
...
+      this, completions_.size(), is_grpc_call_finished_);
+
   MaybeUnregister();
-  if (completions_.empty()) {
-    // Nothing to cancel -- either the call was already finished, or it has
-    // never been started.
-    return;
+
+  if (!completions_.empty() && !is_grpc_call_finished_) {
+    // Important: since the stream always has a pending read operation,
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/grpc_stream.cc]: This comment is hard to parse because:
  * it conflates pre- and post-conditions
  * doesn't clearly indicate normal and error operation
  * uses ambiguous phrasing (e.g. "cancellation" could seemingly apply to either the context or the call)

Consider phrasing something like:

> Important: during normal operation, the stream always has a pending read operation so `Shutdown` would hang indefinitely if we didn't cancel the context_. However, if the stream has already failed, avoid canceling the context to avoid overwriting the status captured during the `OnOperationFailed`.

File: Firestore/core/src/firebase/firestore/remote/grpc_stream.cc [hunk 3]
```diff
@@ -173,22 +179,28 @@ void GrpcStream::FinishAndNotify(const Status& status) {
...
+      "GrpcStream('%s'): shutting down; completion(s): %s, is finished: %s",
+      this, completions_.size(), is_grpc_call_finished_);
+
   MaybeUnregister();
-  if (completions_.empty()) {
-    // Nothing to cancel -- either the call was already finished, or it has
-    // never been started.
-    return;
+
+  if (!completions_.empty() && !is_grpc_call_finished_) {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/grpc_stream.cc]: I don't entirely understand why both conditions must hold here. It seems as if `is_grpc_call_finished_` should be sufficient, no?

In particular this suggests that it's possible for e.g. `completions_` to be empty but `is_grpc_call_finished_` to be false and we wouldn't want to call `FinishGrpcCall`, but that seems wrong.

Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/grpc_stream.cc]: > this suggests that it's possible for e.g. completions_ to be empty but is_grpc_call_finished_ to be false and we wouldn't want to call FinishGrpcCall, but that seems wrong.

The stream is in this state before it is started. Finishing a non-started stream would also fail a gRPC assertion. I don't think we should ever run into the case when `Stream` tries to `Finish` a `GrpcStream` before it is started. I'm mainly doing this because it's an easy check to make.

If you think protecting against this case isn't worthwhile, I'll simplify the check. Otherwise, I'll add a comment.

File: Firestore/core/src/firebase/firestore/remote/grpc_stream.cc (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/grpc_stream.cc]: Thanks, I like this wording.

File: Firestore/core/src/firebase/firestore/remote/grpc_stream.cc (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/grpc_stream.cc]: OK. Looking at the prior state more closely there was a comment there that explained this and it made sense. It's possible you may want to merge something of that comment with what you have below.
